### ISSUE TITLE
Make c library thread-safe

### ISSUE THREAD

**Issue by robgjansen** [igraph/igraph#812]
**Tl;dr** - I would like the C library to provide **thread safety**, i.e., such that igraph does not use modify any global variables simultaneously across different functions.

How much work would this be? I imagine an initialization function that mallocs some memory, which is then used in all library calls later so that the library gets any global state from the initial malloced data instead of from global variables. This approach would mean modifying all library functions to accept this state parameter as input. Alternatively, you could do that internally, and store the data as a thread-local-storage variable, so that each thread initializes its own state.

I think the benefits are clear, but I'll try to illuminate them by explaining my situation below.

As [I've mentioned before](https://github.com/igraph/igraph/issues/728#issuecomment-59152943), I use the igraph C library extensively in my scalable discrete-event network simulator ([Shadow](https://github.com/shadow/shadow)). Thank you for continuing work on the C library, as it is very useful for my simulator.

Shadow is unique in that its "virtual nodes" run real software, but the nodes communicate over a simulated network. Because of this, each Shadow "virtual node" needs its own state, etc. Further, Shadow uses multiple worker threads, and each of them are assigned a number of virtual nodes that they manage. This means that my worker threads may all access igraph at the same time, and usually this ends badly, e.g., in a segfault or in corrupted state.

I use igraph for all nodes to manage the Internet model between them, and I also use it to manage action-dependency graphs for my traffic generator nodes. I currently must lock all accesses to igraph, which harms execution time in my parallel simulator.

Am I missing something, or is thread safety an outstanding issue?

**Comment by ntamas** [comment]
It has not been tested thoroughly yet, but igraph can be compiled with thread-local storage enabled so all the global state information (which is mostly things like a stack where we push pointers that are to be freed in case a function is about to return with an error code) is stored in thread-local variables instead of global ones. E.g., take a look at lines 33-35 in `igraph_error.c`:

```
static IGRAPH_THREAD_LOCAL igraph_error_handler_t *igraph_i_error_handler=0;
static IGRAPH_THREAD_LOCAL char igraph_i_errormsg_buffer[500];
static IGRAPH_THREAD_LOCAL char igraph_i_warningmsg_buffer[500];
```

`IGRAPH_THREAD_LOCAL` is a macro that is replaced by an empty string by default; however, if you enable thread-local storage, it is replaced with `__thread` or `__declspec(thread)`, depending on which format your compiler supports. Theoretically, all the global variables we have should be marked by `IGRAPH_THREAD_LOCAL`, so it should be enough to run `./configure --enable-tls && make` to compile a version of igraph that comes with TLS support. (There is a catch, though: we cannot guarantee that third-party libraries that igraph uses are also thread-safe, so you must compile igraph with our internal BLAS, LAPACK and ARPACK code to ensure that those parts are thread-safe as well).

I have to stress that this has not been tested thoroughly, so it may happen that we did not get this 100% right and there are still race conditions, deadlocks or memory corruption issues. However, it's a first step that had to be taken in order to start adding thread-safety without breaking the existing API. (Otherwise we would have had to pass an "igraph context" variable to all igraph calls and of course create a different context for different threads).

**Comment by robgjansen** [comment]
Ah, a great start! Thanks for the details! I'll give this a try soon.

Is `IGRAPH_THREAD_LOCAL` exported, so that I can use something like the following from my client code to test whether TLS support is built in?

```
#ifdef IGRAPH_THREAD_LOCAL
//do thread-unsafe things
#else
//lock
//do thread-unsafe things
//unlock
#endif
```

**Comment by gaborcsardi** [comment]
Yes, this is what we have:

```c
/** 
 * \define IGRAPH_THREAD_SAFE
 * 
 * Macro that is defined to be 1 if the current build of the 
 * igraph library is thread-safe, and 0 if it is not.
 */

#define IGRAPH_THREAD_SAFE 1
```

I'll close this issue know, but please open a new one if you run into any trouble.

**Comment by robgjansen** [comment]
I assume `IGRAPH_THREAD_SAFE` refers only to igraph itself. Is this correct? What is the method to detect if igraph's internal BLAS, LAPACK and ARPACK code is being used, so we know for sure that using igraph will be fully thread safe?

**Comment by ntamas** [comment]
Yes, `IGRAPH_THREAD_SAFE` refers only to igraph itself. We do not have similar variables that would define whether igraph and all its dependencies are thread-safe, mostly because the answer in most cases would be "no": [GLPK is not thread safe](https://en.wikibooks.org/wiki/GLPK/Using_the_GLPK_callable_library#Thread_safety), and GMP is [thread-safe with some exceptions only](https://gmplib.org/manual/Reentrancy.html).
